File: c_files/raylib/flappy_bird.c
#include "raylib.h"
#include "utils.h"
#include <stdio.h>

#define SCORE_LENGTH 10
#define PILLARS_LENGTH 20

#define VERTICAL_GAP 150
#define HORIZONTAL_GAP 250

// ***********************************
// PLAYER
// ***********************************

typedef struct Player
{
    Rectangle rect;
    Color color;
    float change_y;
    float gravity;
    float jump_strength;

} Player;

void player_init(Player *player)
{
    player->rect = (Rectangle){150, 150, 40, 40};
    player->color = RED;
    player->change_y = 0.0f;
    player->gravity = 1.0f;
    player->jump_strength = -12.0f;
}

void player_draw(Player *player)
{
    DrawRectangleRec(player->rect, player->color);
}

void player_update(Player *player)
{
    if (IsKeyPressed(KEY_UP))
    {
        player->change_y = player->jump_strength;
    }

    player->change_y += player->gravity;
    player->rect.y += player->change_y;

    // player bounds
    player->rect.y = max_float(0.0f,
                               min_float(player->rect.y,
                                         GetScreenHeight() - player->rect.height));
}

// ***********************************
// PILLARS
// ***********************************

typedef struct Pillar
{
    Rectangle rect;
    Color color;
    float speed;
    bool scored;

} Pillar;

void pillar_init(Pillar *pillar)
{
    pillar->rect = (Rectangle){GetScreenWidth(),
                               0,
                               50,
                               GetRandomValue(0, GetScreenHeight() - VERTICAL_GAP)};
    pillar->color = DARKGREEN;
    pillar->speed = 300.0f;
    pillar->scored = false;
}

void pillar_draw(Pillar *pillar)
{
    // upper pillar
    DrawRectangleRec(pillar->rect, pillar->color);

    // lower pillar
    float lower_y = pillar->rect.height + VERTICAL_GAP;
    DrawRectangleRec((Rectangle){pillar->rect.x,
                                 lower_y,
                                 pillar->rect.width,
                                 GetScreenHeight() - lower_y},
                     pillar->color);
}

void pillar_update(Pillar *pillar)
{
    pillar->rect.x -= pillar->speed * GetFrameTime();
}

// PILLARS LIST

typedef struct Pillars
{
    Pillar list[PILLARS_LENGTH];
    int size;

} Pillars;

void pillars_init(Pillars *pillars)
{
    pillars->size = 0;
}

void pillars_draw(Pillars *pillars)
{
    for (int i = 0; i < pillars->size; i++)
    {
        pillar_draw(&pillars->list[i]);
    }
}

void pillars_update(Pillars *pillars)
{
    // update pillar list
    if (pillars->size <= 0 ||
        (pillars->list[pillars->size - 1].rect.x +
         pillars->list[pillars->size - 1].rect.width +
         HORIZONTAL_GAP) <
            GetScreenWidth())
    {
        Pillar pillar;
        pillar_init(&pillar);
        pillars->list[pillars->size] = pillar;
        pillars->size += 1;
    }

    // move pillars
    for (int i = 0; i < pillars->size; i++)
    {
        pillar_update(&pillars->list[i]);

        // remove offscreen pillar
        if (pillars->list[i].rect.x + pillars->list[i].rect.width + HORIZONTAL_GAP < 0)
        {
            for (int j = i; j < pillars->size; j++)
            {
                pillars->list[j] = pillars->list[j + 1];
            }

            pillars->size -= 1;
        }
    }
}

// ***********************************
// GAME MANAGER
// ***********************************

typedef struct Game
{
    int score;
    char score_str[SCORE_LENGTH];
    bool game_over;

    Player player;
    Pillars pillars;

} Game;

void game_init(Game *game)
{
    game->score = 0;
    game->game_over = false;

    player_init(&game->player);
    pillars_init(&game->pillars);
}

void game_draw(Game *game)
{
    player_draw(&game->player);
    pillars_draw(&game->pillars);

    // draw score
    DrawText(game->score_str, 20, 20, 40, BLACK);

    if (game->game_over)
    {
        DrawText("GAME OVER",
                 GetScreenWidth() / 2 - 150,
                 GetScreenHeight() / 2 - 100,
                 60,
                 BLACK);
    }
}

void game_update(Game *game)
{
    if (!game->game_over)
    {
        // convert score to string
        snprintf(game->score_str, SCORE_LENGTH, "%d", game->score);

        player_update(&game->player);
        pillars_update(&game->pillars);

        for (int i = 0; i < game->pillars.size; i++)
        {
            // update score
            if (game->player.rect.x >
                (game->pillars.list[i].rect.x + game->pillars.list[i].rect.width))
            {
                if (!game->pillars.list[i].scored)
                {
                    game->score += 1;
                    game->pillars.list[i].scored = true;
                }
            }

            if (CheckCollisionRecs(game->player.rect, game->pillars.list[i].rect))
            {
                game->game_over = true;
            }
            if (CheckCollisionRecs(game->player.rect,
                                   (Rectangle){game->pillars.list[i].rect.x,
                                               game->pillars.list[i].rect.height + VERTICAL_GAP,
                                               game->pillars.list[i].rect.width,
                                               GetScreenHeight() - game->pillars.list[i].rect.height - VERTICAL_GAP}))
            {
                game->game_over = true;
            }
        }

        // player falls down
        if (game->player.rect.y + game->player.rect.height >= GetScreenHeight())
        {
            game->game_over = true;
        }
    }
    else
    {
        if (IsKeyPressed(KEY_ENTER))
        {
            game->score = 0;
            game->game_over = false;

            player_init(&game->player);
            pillars_init(&game->pillars);
        }
    }
}

// ***********************************
// MAIN
// ***********************************

int main(void)
{
    const int SCREEN_WIDTH = 800;
    const int SCREEN_HEIGHT = 600;
    const char SCREEN_TITLE[] = "Flappy Bird";
    const Color SCREEN_BACKGROUND = SKYBLUE;

    InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

    Game game;
    game_init(&game);

    while (!WindowShouldClose())
    {
        BeginDrawing();
        ClearBackground(SCREEN_BACKGROUND);

        game_draw(&game);
        game_update(&game);

        EndDrawing();
    }

    CloseWindow();

    return 0;
}

File: python/raylib/flappy_bird.py
import pyray as p
from dataclasses import dataclass

SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600
SCREEN_TITLE = "Flappy Bird"
SCREEN_BACKGROUND = p.RAYWHITE


class Player:
    def __init__(self):
        self.x = 200
        self.y = p.get_screen_height() / 2
        self.width = 50
        self.height = 50
        self.color = p.RED
        self.change_y = 0
        self.gravity = 1500
        self.jump_strength = -500
        self.active = False

    def draw(self):
        p.draw_rectangle_rec(
            p.Rectangle(self.x, self.y, self.width, self.height), self.color
        )

    def update(self):
        # jump
        if p.is_key_pressed(p.KeyboardKey.KEY_UP):
            self.change_y = self.jump_strength
            self.active = True

        # add gravity
        if self.active:
            self.change_y += self.gravity * p.get_frame_time()
            self.y += self.change_y * p.get_frame_time()

        # player bounds
        self.y = max(0, min(self.y, p.get_screen_height() - self.height))


@dataclass
class Pillar:
    x: float
    y: float
    width: float
    height: float
    scored: bool


class Pillars:
    def __init__(self):
        self.list = []
        self.vertical_gap = 200
        self.horizontal_gap = 300
        self.color = p.DARKGREEN
        self.speed = 200

    def draw(self):
        for pillar in self.list:
            # top pillar
            p.draw_rectangle_rec(
                p.Rectangle(pillar.x, pillar.y, pillar.width, pillar.height), self.color
            )

            # bottom pillar
            bottom_y = pillar.height + self.vertical_gap
            p.draw_rectangle_rec(
                p.Rectangle(
                    pillar.x,
                    bottom_y,
                    pillar.width,
                    p.get_screen_height() - bottom_y,
                ),
                self.color,
            )

    def update(self):
        # fill pillar list
        if (
            len(self.list) <= 0
            or self.list[-1].x + self.list[-1].width + self.vertical_gap
            < p.get_screen_width()
        ):
            pillar = Pillar(
                p.get_screen_width(),
                0,
                50,
                p.get_random_value(0, p.get_screen_height() - self.horizontal_gap),
                False,
            )

            self.list.append(pillar)

        # move and remove pillars
        for pillar in self.list:
            pillar.x -= self.speed * p.get_frame_time()

            # remove offscreen pillars
            if pillar.x + pillar.width + self.horizontal_gap < 0:
                self.list.remove(pillar)


class Game:
    def __init__(self):
        self.score = 0
        self.game_over = False

        self.pillars = Pillars()
        self.player = Player()

    def draw(self):
        self.pillars.draw()
        self.player.draw()

        # draw score
        p.draw_text(f"Score: {self.score}", 10, 10, 30, p.BLACK)

        if self.game_over:
            p.draw_text(
                "GAME OVER",
                p.get_screen_width() // 2 - 100,
                p.get_screen_height() // 2 - 20,
                40,
                p.BLACK,
            )
            p.draw_text(
                "Press ENTER to restart",
                p.get_screen_width() // 2 - 150,
                p.get_screen_height() // 2 + 100,
                30,
                p.BLACK,
            )

    def update(self):
        if not self.game_over:
            self.player.update()
            self.pillars.update()

            # pillars collision player
            for pillar in self.pillars.list:
                # update scores
                if self.player.x > pillar.x + pillar.width:
                    if not pillar.scored:
                        self.score += 1
                        pillar.scored = True

                if p.check_collision_recs(
                    p.Rectangle(pillar.x, pillar.y, pillar.width, pillar.height),
                    p.Rectangle(
                        self.player.x,
                        self.player.y,
                        self.player.width,
                        self.player.height,
                    ),
                ):
                    self.game_over = True

                bottom_y = pillar.height + self.pillars.vertical_gap
                if p.check_collision_recs(
                    p.Rectangle(
                        pillar.x,
                        bottom_y,
                        pillar.width,
                        p.get_screen_height() - bottom_y,
                    ),
                    p.Rectangle(
                        self.player.x,
                        self.player.y,
                        self.player.width,
                        self.player.height,
                    ),
                ):
                    self.game_over = True

            # player falling down
            if self.player.y >= p.get_screen_height() - self.player.height:
                self.game_over = True

        else:
            if p.is_key_pressed(p.KeyboardKey.KEY_ENTER):
                self.game_over = False
                self.score = 0

                self.pillars = Pillars()
                self.player = Player()


def main():
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    game = Game()

    while not p.window_should_close():
        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        game.draw()
        game.update()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()
